cmd/provisioner-localpv/app/provisioner.go

```diff
@@ -0,0 +1,205 @@
+)
+
+const (
+	KeyNode = "kubernetes.io/hostname"
+)
+
+// NewProvisioner will create a new Provisioner object and initialize
+//  it with global information used across PV create and delete operations.
+func NewProvisioner(stopCh chan struct{}, kubeClient *clientset.Clientset) (*Provisioner, error) {
+
+	namespace := getOpenEBSNamespace() //menv.Get(menv.OpenEBSNamespace)
+	if len(strings.TrimSpace(namespace)) == 0 {
+		return nil, fmt.Errorf("Cannot start Provisioner: failed to get namespace")
+	}
+
+	p := &Provisioner{
+		stopCh: stopCh,
+
+		kubeClient:  kubeClient,
+		namespace:   namespace,
+		helperImage: getDefaultHelperImage(),
+		defaultConfig: []mconfig.Config{
+			{
+				Name:  KeyPVBasePath,
+				Value: getDefaultBasePath(),
+			},
+		},
+	}
+	p.configParser = p.CASConfigParser
+
+	return p, nil
+}
+
+// SupportsBlock will be used by controller to determine if block mode is
+//  supported by the host path provisioner. Return false.
+func (p *Provisioner) SupportsBlock() bool {
+	return false
+}
+
+// Provision is invoked by the PVC controller which expect the PV
+//  to be provisioned and a valid PV spec returned.
+func (p *Provisioner) Provision(opts pvController.VolumeOptions) (*v1.PersistentVolume, error) {
+	pvc := opts.PVC
+	if pvc.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim.Spec.Selector is not supported")
+	}
+	for _, accessMode := range pvc.Spec.AccessModes {
+		if accessMode != v1.ReadWriteOnce {
+			return nil, fmt.Errorf("Only support ReadWriteOnce access mode")
+		}
+	}
+	node := opts.SelectedNode
+	if opts.SelectedNode == nil {
+		return nil, fmt.Errorf("configuration error, no node was specified")
+	}
+
+	name := opts.PVName
+
+	// Create a new Config instance for the PVC that will help with
+	// generating a valid host path to be used by the PV.
+	pvcConfig, err := p.configParser(name, pvc)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO: Determine if hostpath or device based Local PV should be created
+	path, err := pvcConfig.GetPath()
+	if err != nil {
+		return nil, err
+	}
+
+	glog.Infof("Creating volume %v at %v:%v", name, node.Name, path)
+
+	// VolumeMode will always be specified as Filesystem for host path volume,
+	// and the value passed in from the PVC spec will be ignored.
+	fs := v1.PersistentVolumeFilesystem
+
+	// It is possible that the HostPath doesn't already exist on the node.
+	// Set the Local PV to create it.
+	hostPathType := v1.HostPathDirectoryOrCreate
+
+	// TODO initialize the Labels and annotations
+	// Use annotations to specify the context using which the PV was created.
+	//volAnnotations := make(map[string]string)
+	//volAnnotations[string(v1alpha1.CASTypeKey)] = casVolume.Spec.CasType
+	//fstype := casVolume.Spec.FSType
+
+	//labels := make(map[string]string)
+	//labels[string(v1alpha1.CASTypeKey)] = casVolume.Spec.CasType
+	//labels[string(v1alpha1.StorageClassKey)] = *className
+
+	//TODO Change the following to a builder pattern
+	return &v1.PersistentVolume{
```

Comment:
This should make use of builder pattern as per idiomatic maya standard.

Reply:
done.